go/tools/builders/compilepkg.go

```diff
@@ -383,16 +386,24 @@ func compileArchive(
 		}
 	}
 
+	pkgDefFile := filepath.Join(filepath.Dir(outPath), pkgDef)
```

Comment:
Move this code and the calls to `appendFile` into a separate block below the `// Check results from nogo` block.

Add a comment explaining what we're doing:

```
// Extract the export data file and pack it in an .x archive together with the
// nogo facts file (if there is one). This allows compile actions to depend
// on .x files only, so we don't need to recompile a package when one of its
// imports changes in a way that doesn't affect export data.
// TODO(golang/go#33820): Ideally, we would use -linkobj to tell the compiler
// to create separate .a and .x files for compiled code and export data, then
// copy the nogo facts into the .x file. Unfortunately, when building a plugin,
// the linker needs export data in the .a file. To work around this, we copy
// the export data into the .x file ourselves.
```